### PR TITLE
Removing MSVC C4800 from pragma block at the top of pybind11.h

### PR DESCRIPTION
**Pull request by rwgk**
Follow-on to PR #3127, based on results obtained under PR #3125.

The suggested changelog entry will be in the final PR of this cleanup series.

### CODE REVIEW

File: include/pybind11/cast.h [hunk 1]
```diff
@@ -322,7 +322,7 @@ template <> class type_caster<bool> {
             }
             #endif
             if (res == 0 || res == 1) {
-                value = (bool) res;
```
Comment (henryiii) [inline include/pybind11/cast.h:325]: Why not just make these `value = res != 0`; ? The problem is that (bool) int_value is poor for performance (according to the warning), so we shouldn't be pretending this is okay to cast.

**Comment by Skylion007** [comment]
Actually never mind, I can add that in a separate PR as that check is much more strict than this one even is and will touch a lot more code.

**Review by Skylion007 — APPROVED** [review]

**Comment by rwgk** [comment]
> Actually never mind, I can add that in a separate PR as that check is much more strict than this one even is and will touch a lot more code.

Awesome, I'm glad. Combining (convoluting) things backfires too often. Best to keep things compartmentalized.

File: include/pybind11/cast.h (hunk 1 above)
Comment (rwgk) [reply include/pybind11/cast.h:325]: Done. Applied globally. — I was aiming for a robust general solution before, but you're right, we only need specific solutions, and `!= 0` is simpler therefore better.

**Pull request merged**